### PR TITLE
Glibc validation test now performed also on FreeBSD

### PR DESCRIPTION
**Pull request by uraimo**
#### What's in this pull request?

As discussed in #3276, the glibc validation tests could be performed at least on FreeBSD too, this PR removes the requirement for a Linux OS. 
I didn't build on FreeBSD but I've verified these two tests with a small c program, so both should pass (closing an invalid fd gives EBADF, presence of the TIOCGWINSZ tty ioctl constant).
Couldn't verify it on Android (and that's why it's not included here) since i don't have a device ATM, but both tests should pass.

---

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by modocache** [comment]
Awesome, thanks @uraimo! I'll give it a try on Android and submit a follow-up pull request soon.

@swift-ci please test and merge

**Comment by uraimo** [comment]
Yep that would be way better, the environment would be linked to the test and less error-prone than simple #ifs , this is what i came up with (better than skipping 10 platforms, also because new platform will _not_ be skipped by default):

```
GlibcTestSuite.test("errno") {
  .skip(.custom({ !( 
            linuxAny(reason:"").evaluate() || freeBSDAny(reason:"").evaluate()
   ) }, reason:"")).code {
     errno = 0
     expectEqual(0, errno)
     close(-1)
     expectEqual(EBADF, errno)
 }
```

Did i miss a better way to do this?

**Comment by modocache** [comment]
Maybe `.only()` would work better? Is it possible to write something like:

```swift
GlibcTestSuite.test("errno")
  .only(.custom({
    linuxAny(reason:"").evaluate() || freeBSDAny(reason:"").evaluate()
   )}, reason:"Only Linux and FreeBSD are supported for now")).code {
    // ...
 }
```

**Comment by uraimo** [comment]
No problem, i was about to ask for more info since i couldn't find it anywhere 😄 

Updated!

**Comment by gribozavr** [comment]
I'm afraid as written this test won't compile on Apple platforms... let's see.

@swift-ci Please test

File: validation-test/stdlib/Glibc.swift [hunk 1]
```diff
@@ -1,27 +1,34 @@
...
+
 var GlibcIoctlConstants = TestSuite("GlibcIoctlConstants")
 
-GlibcIoctlConstants.test("tty ioctl constants availability") {
-  let aConstant = TIOCGWINSZ
+GlibcIoctlConstants.test("tty ioctl constants availability") 
+  .skip(.custom({
+    !( linuxAny(reason: "").evaluate() || freeBSDAny(reason: "").evaluate() ) 
+  }, reason: "Only Linux and FreeBSD are supported")).code {
+    let aConstant = TIOCGWINSZ
```
Comment (gribozavr) [inline validation-test/stdlib/Glibc.swift]: Without any imports this constant won't be available on Apple platforms.

**Comment by gribozavr** [comment]
I suggest that we add:

```swift
// UNSUPPORTED: OS=macosx
// UNSUPPORTED: OS=ios
// UNSUPPORTED: OS=tvos
// UNSUPPORTED: OS=watchos
```

and remove all `#if`s and `.skip`s.

File: validation-test/stdlib/Glibc.swift [hunk 2]
```diff
@@ -1,27 +1,34 @@
...
-GlibcTestSuite.test("errno") {
-  errno = 0
-  expectEqual(0, errno)
-  close(-1)
-  expectEqual(EBADF, errno)
+GlibcTestSuite.test("errno")
+  .skip(.custom({
+    !( linuxAny(reason: "").evaluate() || freeBSDAny(reason: "").evaluate() ) 
+  }, reason: "Only Linux and FreeBSD are supported")).code {
+    errno = 0
```
Comment (gribozavr) [inline validation-test/stdlib/Glibc.swift]: Please use 2 spaces for indentation.

**Comment by uraimo** [comment]
@gribozavr is that `import Swift` really necessary in this case?

**Comment by uraimo** [comment]
Compilation failed, I could't compile it locally and as expected if was full of typos :)

**Comment by gribozavr** [comment]
> @gribozavr is that `import Swift` really necessary in this case?

If you change the test to use `%target-run-simple-swift`, it won't be.

**Comment by uraimo** [comment]
Hmm, wait, the CI did the merge even if the tests were failing... preparing a new PR.

**Comment by modocache** [comment]
Swift CI must have merged an earlier commit, because of my "please test and merge" comment. I'm not sure how it behaves when the pull request is modified after the CI comment. I apologize for prompting CI too quickly here; this deserved more review. Sorry! 🙇

**Comment by benlangmuir** [comment]
I reverted this commit in 98829a2 because it broke the OS X bots.

**Comment by modocache** [comment]
Thanks @benlangmuir! Apologies again, I'll be more careful in the future.

This was totally my fault, but: @shahmishal, I think it should be impossible for "please test and merge" to merge commits that break CI. I requested a "test and merge" for an earlier version of this pull request, but CI merged c145ac2424ea95291aedec2c38b58e7344075ce3, an updated version that didn't pass the tests.

CI bots at Facebook cancel any jobs (be they test or merge jobs) when updates are detected in the patch set. Would it make sense to do the same for Swift CI?

**Comment by shahmishal** [comment]
@modocache Sounds good, I will look into this. Thanks!

**Comment by uraimo** [comment]
Next time I'll try to complete a local build as usual before pushing changes, thanks to @benlangmuir that reverted it promptly. +1

**Pull request merged**